[`perflint`] Implement fix for `manual-dict-comprehension` (`PERF403`)

w0nder1ng opened the pull request:
## Summary

This change adds an auto-fix for manual dict comprehensions. It also copies many of the improvements from #13919 (and associated PRs fixing issues with it), and moves some of the utility functions from `manual_list_comprehension.rs` into a separate `helpers.rs` to be used in both. 

## Test Plan

I added a preview test case to showcase the new fix and added a test case in `PERF403.py` to make sure lines with semicolons function. I didn't yet make similar tests to the ones I added earlier to `PERF401.py`, but the logic is the same, so it might be good to add those to make sure they work.

Skylion007 commented:
Slight nit: had a situation where it didn't simplify an `annotated dict = dict()` builtin to a single line comprehension. Still kept a sepereate call to the `dict()` builtin function followed by update. Manually fixed that.

w0nder1ng commented:
I had forgotten about using a function call to make a dict, since it's not typically something I do in my own code. I've now modified the "is empty dict" check to allow builtin function calls for both PERF403 and PERF401.

Skylion007 commented:
Title is wrong, it's PERF403, right?

w0nder1ng commented:
Do you know why the tests aren't running? The test jobs in my fork of the repo are timing out while waiting for a runner.

ntBre commented:
I think the most recent commits came in while CI was disabled. You can try pushing another (possibly empty) commit, or closing and reopening the PR to retrigger CI.

Skylion007 commented:
@ntBre CI is run now.

ntBre commented:
Thanks for the ping, I'll try to review this tomorrow.

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+        ast::Stmt::Assign(assign) => assign.targets.len() == 1,
+        _ => false,
+    });
+    // If the binding gets used in between the assignment and the for loop, a list comprehension is no longer safe
```
```suggestion
    // If the binding gets used in between the assignment and the for loop, a dict comprehension is no longer safe
```

or just `comprehension` maybe?

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+    });
+    // If the binding gets used in between the assignment and the for loop, a list comprehension is no longer safe
+
+    // If the binding is after the for loop, then it can't be fixed, and this check would panic,
+    // so we check that they are in the same statement first
+    let binding_unused_between = assignment_in_same_statement
+        && binding_stmt.is_some_and(|binding_stmt| {
+            let from_assign_to_loop = TextRange::new(binding_stmt.end(), for_stmt.start());
+            // Test if there's any reference to the list symbol between its definition and the for loop.
+            // if there's at least one, then it's been accessed in the middle somewhere, so it's not safe to change into a list comprehension
+            !binding
+                .references()
+                .map(|ref_id| checker.semantic().reference(ref_id).range())
+                .any(|text_range| from_assign_to_loop.contains_range(text_range))
+        });
```
I thought we already checked the end of the `for` loop vs any references to the variable above? Maybe I'm missing something.

w0nder1ng replied:
This checks whether the result dictionary is accessed between its assignment and the for loop. I was trying to get across that it's not safe to transform this:
```python
res = {}
print(len(res))
values = {"a": 1, "b": 2}
for k, v in values.items():
    res[k] = v
```
into this:
```python
print(len(res))  # NameError
values = {"a": 1, "b": 2}
res = {k: v for k, v in values.items()}
```
In this case, it needs to be `.update` instead, since the binding statement needs to stay put

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+        *range,
+    );
+
+    if checker.settings.preview.is_enabled() {
```
How hard/intrusive would it be to check this earlier? There's a lot of new code above this that would be nice to skip if we're not going to use the fix. However, it's also nice not to have to change much when stabilizing the rule, so if it's more than moving this check up a bit, don't worry about it.

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+
+    if checker.settings.preview.is_enabled() {
+        // k: v inside the comprehension
+        let to_append = (slice.as_ref(), value.as_ref());
```
I'd lean towards passing these as separate `key` and `value` arguments. If this was to avoid `clippy::too_many_arguments`, it's fine to add an `#[allow(...)]` here.

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -42,17 +46,45 @@ use crate::checkers::ast::Checker;
 
 /// PERF403
-pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body: &[Stmt]) {
+pub(crate) fn manual_dict_comprehension(checker: &mut Checker, for_stmt: &ast::StmtFor) {
```
Does this need to be `&mut`? With `report_diagnostic`, I think `&` should be fine.

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
````diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+    //      ...
+    // ```
+    // becomes
+    // [... for i in (a, b)]
````
```suggestion
    // {... for i in (a, b)}
```

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
````diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+    // ```
+    // becomes
+    // [... for i in (a, b)]
+    let iter_str = if for_stmt
+        .iter
+        .as_ref()
+        .as_tuple_expr()
+        .is_some_and(|expr| !expr.parenthesized)
````
This is slightly shorter but maybe not more readable. Up to you.
```suggestion
    let iter_str = if let Expr::Tuple(ast::ExprTuple {
        parenthesized: false,
        ..
    }) = &*for_stmt.iter
```

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+        ],
+    );
+
+    let newline = checker.stylist().line_ending().as_str();
+
+    let indent = locator.slice(TextRange::new(
+        locator.line_start(for_stmt.range.start()),
+        for_stmt.range.start(),
+    ));
```
This is all a nice touch. I don't usually worry about formatting when generating a fix :smile:

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -163,6 +228,234 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+
+    match fix_type {
+        DictComprehensionType::Update => {
+            let variable_name = locator.slice(binding);
+            let indentation = if for_loop_inline_comments.is_empty() {
+                String::new()
+            } else {
+                format!("{newline}{indent}")
+            };
```
It looks like these two are used in both arms of the `match` and could be hoisted out.

w0nder1ng replied:
The indentation depends on whether the comments are empty, and the relevant comments depend on the type of fix, so I think indentation has to stay in both.

ntBre replied:
Oh I see now, thanks.

ntBre commented on crates/ruff_linter/src/rules/perflint/snapshots/ruff_linter__rules__perflint__tests__preview__PERF403_PERF403.py.snap:
```diff
@@ -0,0 +1,204 @@
+100     |-        # comment 1
+101     |-    }
+102     |-    for idx, name in enumerate(
+    99  |+    # comment 3
+    100 |+    # comment 4
+    101 |+    # comment 1
```
This is kind of a strange order. Is this how the list version works? It's already an unsafe fix, so it's pretty reasonable to drop the comments instead of trying to sort them out, if it's difficult to do. We just need to mention it in the `## Fix safety` docs.

ntBre reviewed (commented):
Thanks for all of this work! It looks good, and most of my inline comments are just stylistic nits.

For tests, I think I'd like to see ~1 test for each of the (very helpful) comments that document some constraint, if that's not too hard. Is that what you had in mind in the PR summary?

w0nder1ng commented on crates/ruff_linter/src/rules/perflint/helpers.rs:
Yes, there are no changes. I just extracted them into functions so I could reuse the code in `manual_dict_comprehension.rs`.

w0nder1ng commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -115,14 +147,47 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+                    .bindings
+                    .iter()
+                    .find(|binding| target.range() == binding.range)
+                    .expect("for-loop target binding must exist");
```
I was thinking that if it doesn't exist, then there is some flaw in my code, so it'd be better to find out with a panic than to silently ignore the logic error. Is there a better way to report an invariant being broken without a panic?

w0nder1ng commented:
@ntBre I think I've addressed all of your feedback, so please let me know if you need anything else changed. 

Also, do you know why the `configuration.md` output test is failing? I don't think I touched any files that affect it.

ntBre commented:
> Also, do you know why the `configuration.md` output test is failing? I don't think I touched any files that affect it.

At the very top of the error message it says 

> Error: docs/configuration.md changed, please run `cargo dev generate-all`:

I think adding a fix probably changed the docs slightly, and you just need to run `cargo dev generate-all` to fix it. Some of the diffs look kind of weird, though, so you may also need to merge `main` or something, but try `generate-all` first.

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -164,5 +221,233 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, target: &Expr, body:
+            DictComprehensionType::Update
+        };
+
+        let mut diagnostic = Diagnostic::new(
+            ManualDictComprehension {
+                fix_type,
+                is_async: for_stmt.is_async,
+            },
+            *range,
+        );
+
+        diagnostic.set_fix(convert_to_dict_comprehension(
+            fix_type,
+            binding,
+            for_stmt,
+            if_test.map(std::convert::AsRef::as_ref),
+            slice.as_ref(),
+            value.as_ref(),
+            checker,
+        ));
```
You can use `Diagnostic::with_fix` here

ntBre reviewed (approved):
Thanks! I'd still like to avoid the `expect` calls, and I added one more suggestion for `Diagnostic::with_fix` but this looks good to me otherwise, once the tests are passing.

ntBre commented:
> Turns out I just needed to merge with main. Did you also want the `.expect()` in `convert_to_dict_comprehension` removed?

Yes please!

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_dict_comprehension.rs:
```diff
@@ -139,8 +139,13 @@ pub(crate) fn manual_dict_comprehension(checker: &Checker, for_stmt: &ast::StmtF
-            .find(|binding| target.range() == binding.range)
-            .expect("for-loop target binding must exist");
+            .find(|binding| target.range() == binding.range);
+        debug_assert!(target_binding.is_some());
```
You could include your previous `expect` message as the `debug_assert` message, if you want.
```suggestion
        debug_assert!(target_binding.is_some(), "for-loop target binding must exist");
```

w0nder1ng commented:
@ntBre let me know if any other changes are necessary

ntBre commented on crates/ruff_linter/src/rules/perflint/rules/manual_list_comprehension.rs:
```diff
@@ -264,12 +263,23 @@ pub(crate) fn manual_list_comprehension(checker: &Checker, for_stmt: &ast::StmtF
-        list_binding_value.is_some_and(|binding_value| match binding_value.as_list_expr() {
-            Some(list_expr) => list_expr.elts.is_empty(),
-            None => false,
+        list_binding_value.is_some_and(|binding_value| match binding_value {
+            // `value = []`
+            Expr::List(list_expr) => list_expr.is_empty(),
+            // `value = list()`
+            // This is probably be linted against, but turning it into a list comprehension will also remove it
+            Expr::Call(call) => {
+                checker
+                    .semantic()
+                    .resolve_builtin_symbol(&call.func)
+                    .is_some_and(|name| name == "list")
+                    && call.arguments.is_empty()
+            }
+            _ => false,
```
Sorry for not noticing this earlier (or for bringing it up again if we already discussed it), but should this be included here? I think these changes (and the corresponding ones on the dict side) are causing the differences in the ecosystem check, and I would kind of like to limit this PR strictly to the autofix.

I would have at least expected some PERF401 snapshots to change with this, since it's obviously affecting ecosystem packages.

Skylion007 replied:
We can split it off in a different PR, but it's an ecosystem bug we noticed when testing.

ntBre commented:
Thanks, I thought we had to revert the `Call` check in the `dict` too, but that's in the new preview part anyway.

Skylion007 commented:
@w0ndering Mind opening a new PR for that ecosystem fix to list Calls as well?
